cluster-autoscaler/simulator/scheduler_based_predicates_checker.go

```diff
@@ -0,0 +1,339 @@
+		// Be sure that the node is schedulable.
+		if nodeInfo.Node().Spec.Unschedulable {
+			continue
+		}
+
+		filterStatuses := p.framework.RunFilterPlugins(context.TODO(), state, pod, nodeInfo)
+		ok := true
+		for _, filterStatus := range filterStatuses {
+			if !filterStatus.IsSuccess() {
+				ok = false
+				break
+			}
+		}
+		if ok {
+			return nodeInfo.Node().Name, nil
+		}
+	}
+	return "", fmt.Errorf("cannot put pod %s on any node", pod.Name)
+}
+
+// CheckPredicates checks if the given pod can be placed on the given node.
+func (p *SchedulerBasedPredicateChecker) CheckPredicates(clusterSnapshot ClusterSnapshot, pod *apiv1.Pod, nodeInfo *scheduler_nodeinfo.NodeInfo) PredicateError {
+	if clusterSnapshot != nil {
+		return p.checkPredicates(clusterSnapshot, pod, nodeInfo.Node().Name)
+	}
+	return p.checkPredicatesDeprecated(pod, nodeInfo)
+}
+
+func (p *SchedulerBasedPredicateChecker) checkPredicatesDeprecated(pod *apiv1.Pod, nodeInfo *scheduler_nodeinfo.NodeInfo) PredicateError {
+	p.delegatingSharedLister.UpdateDelegate(p.informerBasedShardLister)
+	defer p.delegatingSharedLister.ResetDelegate()
+
+	state := scheduler_framework.NewCycleState()
+	preFilterStatus := p.framework.RunPreFilterPlugins(context.TODO(), state, pod)
+	if !preFilterStatus.IsSuccess() {
+		return NewPredicateError(
+			InternalPredicateError,
+			"",
+			preFilterStatus.Message(),
+			preFilterStatus.Reasons(),
+			emptyString)
+	}
+
+	filterStatuses := p.framework.RunFilterPlugins(context.TODO(), state, pod, nodeInfo)
+	for filterName, filterStatus := range filterStatuses {
+		if !filterStatus.IsSuccess() {
+			if filterStatus.IsUnschedulable() {
+				return NewPredicateError(
+					NotSchedulablePredicateError,
+					filterName,
+					filterStatus.Message(),
+					filterStatus.Reasons(),
+					p.buildDebugInfo(filterName, nodeInfo))
+			}
+			return NewPredicateError(
+				InternalPredicateError,
+				filterName,
+				filterStatus.Message(),
+				filterStatus.Reasons(),
+				p.buildDebugInfo(filterName, nodeInfo))
+		}
+	}
+	return nil
+}
+
+func (p *SchedulerBasedPredicateChecker) checkPredicates(clusterSnapshot ClusterSnapshot, pod *apiv1.Pod, nodeName string) PredicateError {
+	schedulerLister, err := clusterSnapshot.GetSchedulerLister()
+	if err != nil {
+		// // TODO(scheduler_framework_integration) distinguish from internal error and predicate error
+		klog.Errorf("Error obtaining SharedLister from clusterSnapshot ")
+		return GenericPredicateError()
+	}
+
+	p.delegatingSharedLister.UpdateDelegate(schedulerLister)
+	defer p.delegatingSharedLister.ResetDelegate()
+
+	state := scheduler_framework.NewCycleState()
+	preFilterStatus := p.framework.RunPreFilterPlugins(context.TODO(), state, pod)
+	if !preFilterStatus.IsSuccess() {
+		return NewPredicateError(
+			InternalPredicateError,
+			"",
+			preFilterStatus.Message(),
+			preFilterStatus.Reasons(),
+			emptyString)
+	}
+
+	nodeInfo, err := schedulerLister.NodeInfos().Get(nodeName)
```

Comment:
Maybe do that before running prefilters, so we fail quickly if nodeInfo doesn't exist in snapshot?